File: python/RemoveKDigits402.py
# Given string num representing a non-negative integer num, and an integer k, return the smallest possible integer
# after removing k digits from num.
def removeKdigits(num: str, k: int) -> str:
    stack = []
    for digit in num:
        while stack and stack[-1] > digit and k > 0:
            k -= 1
            stack.pop()
        stack.append(digit)
    while k > 0:
        k -= 1
        stack.pop()
    return ''.join(stack).lstrip('0') or '0'


if __name__ == "__main__":
    print(removeKdigits("1432219", 3))  # 1219
    print(removeKdigits("10200", 1))  # 200
    print(removeKdigits("10", 2))  # 0
    print(removeKdigits("9", 1))  # 0
    print(removeKdigits("112", 1))  # 11
    print(removeKdigits("1234567890", 9))  # 0
    print(removeKdigits("1234567890", 10))  # 0

File: java/RemoveKDigits402.java
import java.util.ArrayDeque;

/*
Given string num representing a non-negative integer num, and an integer k, return the smallest possible integer after removing k digits from num.
*/
public class RemoveKDigits402 {
    public static void main(String[] args) {
        System.out.println(removeKdigits("1432219", 3));
        System.out.println(removeKdigits("10200", 1));
        System.out.println(removeKdigits("10", 2));
        System.out.println(removeKdigits("9", 1));
    }

    public static String removeKdigits(String num, int k) {
        ArrayDeque<Character> stack = new ArrayDeque<>();
        for (char c : num.toCharArray()) {
            while (!stack.isEmpty() && stack.peekLast() > c && k > 0) {
                k--;
                stack.removeLast();
            }
            stack.addLast(c);
        }
        while (!stack.isEmpty() && k > 0) {
            k--;
            stack.removeLast();
        }
        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) sb.append(stack.removeFirst());
        while (sb.length() > 1 && sb.charAt(0) == '0') sb.deleteCharAt(0);
        if (sb.isEmpty()) return "0";
        return sb.toString();
    }
}
